### PR TITLE
Added support for CTE without columns

### PR DESCRIPTION
**Pull request by soham-bentley**
Fixes https://github.com/iTwin/itwinjs-backlog/issues/814
Fixes https://github.com/iTwin/itwinjs-backlog/issues/1226

### CODE REVIEW

File: iModelCore/ECDb/Tests/NonPublished/ECSqlStatementTests.cpp [hunk 1]
```diff
@@ -11619,10 +11619,7 @@ TEST_F(ECSqlStatementTestFixture, SelectAnySomeAll)
...
-        Utf8PrintfString expectedSql ("SELECT [SomeEntity].[ECInstanceId],[SomeEntity].[ECClassId],[SomeEntity].[Name],[SomeEntity].[Primary],[SomeEntity].[Secondary],[SomeEntity].[Random] "
-            "FROM (SELECT [Id] ECInstanceId,%d ECClassId,[Name],[Primary],[Secondary],[Random] FROM [main].[ts_SomeEntity]) [SomeEntity] "
-            "WHERE EXISTS(SELECT [SomeEntity].[Secondary] FROM (SELECT [Id] ECInstanceId,%d ECClassId,[Secondary] FROM [main].[ts_SomeEntity]) [SomeEntity] WHERE [SomeEntity].[Primary] = [Secondary])", classId, classId);
-
+        Utf8PrintfString expectedSql ("SELECT [SomeEntity].[ECInstanceId],[SomeEntity].[ECClassId],[SomeEntity].[Name],[SomeEntity].[Primary],[SomeEntity].[Secondary],[SomeEntity].[Random] FROM (SELECT [Id] ECInstanceId,%d ECClassId,[Name],[Primary],[Secondary],[Random] FROM [main].[ts_SomeEntity]) [SomeEntity] WHERE EXISTS(SELECT [SomeEntity].[Secondary] FROM (SELECT [Id] ECInstanceId,%d ECClassId,[Secondary] FROM [main].[ts_SomeEntity]) [SomeEntity] WHERE [SomeEntity].[Primary] = [SomeEntity].[Secondary])", classId, classId);
```
Comment (ColinKerr) [inline iModelCore/ECDb/Tests/NonPublished/ECSqlStatementTests.cpp:11622]: Is there a reason you switched the formatting on these expected results?

Comment (soham-bentley) [reply iModelCore/ECDb/Tests/NonPublished/ECSqlStatementTests.cpp]: No actually I had to change the expected sql string as per my changes, doing that the formatting also got messed up.
Anyways I have fixed the formatting.

File: iModelCore/ECDb/Tests/NonPublished/CommonTableExpTests.cpp [hunk 2]
```diff
@@ -1594,14 +1594,139 @@ TEST_F(CommonTableExpTestFixture, CTE_Subquery_Tests) {
...
...Id] IS NULL THEN NULL ELSE 38 END) [SchemaRelECClassId],[Name],[DisplayLabel],[Description],[Type],[Modifier],[CustomAttributeContainerType],[RelationshipStrength],[RelationshipStrengthDirection] FROM [main].[ec_Class]) [ECClassDef] WHERE EXISTS(WITH a AS (SELECT [ClassHasBaseClasses].[ECInstanceId] [K0],[ClassHasBaseClasses].[ECClassId] [K1],[ClassHasBaseClasses].[Ordinal] [K2],[ClassHasBaseClasses].[SourceECInstanceId] [K3],[ClassHasBaseClasses].[SourceECClassId] [K4],[ClassHasBaseClasses].[TargetECInstanceId] [K5],[ClassHasBaseClasses].[TargetECClassId] [K6] FROM (SELECT [ec_ClassHasBaseClasses].[Id] [ECInstanceId],42 [ECClassId],[ec_ClassHasBaseClasses].[ClassId] [SourceECInstanceId],37 [SourceECClassId],[ec_ClassHasBaseClasses].[BaseClassId] [TargetECInstanceId],37 [TargetECClassId],[Ordinal] FROM [main].[ec_ClassHasBaseClasses]) [ClassHasBaseClasses])\nSELECT [K0],[K1],[K2],[K3],[K4],[K5],[K6] FROM a WHERE [ECClassDef].[ECInstanceId] >= [K0] OR [ECClassDef].[ECInstanceId] >= [K1] OR [ECClassDef].[ECInstanceId] >= [K2] OR [ECClassDef].[ECInstanceId] >= [K3] OR [ECClassDef].[ECInstanceId] >= [K4] OR [ECClassDef].[ECInstanceId] >= [K5] OR [ECClassDef].[ECInstanceId] >= [K6])" );
```
Comment (rschili) [inline iModelCore/ECDb/Tests/NonPublished/CommonTableExpTests.cpp:1616]: Hey, I while the changes themselves look fine to me, these tests are not good. I see you already pushed similar tests to this file in a previous "Added support for CTE in subquery PR#836".
I know some of our tests check for string equality against the NativeSQL, but it's a bad habit. Just look at these tests. It's near impossible to tell what is wrong if one of them starts failing. The NativeSql depends on so many factors and a slight change in a future story may cause it to look different. Then, somebody has to go and adjust all these expected results by hand.

It's much better to test against the actual results and column infos of the query. It looks to me as if that's possible for these queries. These enormous NativeSql strings are going to be a pain to deal with in the future if we don't adjust it now.

**Review by rschili — CHANGES_REQUESTED** [review]

File: iModelCore/ECDb/Tests/NonPublished/CommonTableExpTests.cpp (hunk 2 above)
Comment (soham-bentley) [reply iModelCore/ECDb/Tests/NonPublished/CommonTableExpTests.cpp:1616]: Sure I will be adding tests against real data as well then tomorrow

**Review by rschili — APPROVED** [review]

**Pull request merged**